alg/crc32c_sse42.c

```diff
@@ -0,0 +1,38 @@
+#include "cpusupport.h"
+#ifdef CPUSUPPORT_X86_CRC32
+
+#include <smmintrin.h>
+
+#include "crc32c_sse42.h"
+
+/**
+ * CRC32C_Update_SSE42(state, buf, len):
+ * Feed ${len} bytes from the buffer ${buf} into the CRC32C whose state is
+ * ${state}.  This implentation uses x86 SSE4.2 instructions, and should only
+ * be used if CPUSUPPORT_X86_CRC32 is defined and cpusupport_x86_crc32 returns
+ * nonzero.
+ */
+void
+CRC32C_Update_SSE42(uint32_t * state, const uint8_t * buf, size_t len)
+{
+	const size_t pre = ((uintptr_t)buf & 7) < len ? (uintptr_t)buf & 7 : len;
+	const size_t remaining = len - pre;
+	const size_t block_len = remaining - remaining % 8;
+	size_t i = 0;
+
+	/* Process bytes before the alignment. */
+	for (; i < pre; i++)
+		*state = _mm_crc32_u8(*state, buf[i]);
+
+	/* Process blocks of 8. */
+	for (; i < block_len; i += 8) {
```

Comment:
... and what will `buf[i]` be when we enter this loop?

Reply:
Well, `0x10003 & 7` is `3`, and `3 < 16`, so `pre=3` and `block_len = 8`.  So when we get here, `i=3`, and `buf[i] = 0x10006` ?  I'm not seeing the problem.

I've rewritten the logic about this, though.  There's no point playing games with `const` vars at the expense of code clarify, especially since this function is only called once per buffer.  If the new code is wrong, at least it should be easier to see why it's wrong.